File: backend/lyrics/lyrics_service.py
import os
import re
import requests
from typing import Dict, Any, Optional
from parser import extract_lyrics

# Simple in-memory cache to prevent spamming APIs and scrapers
cache: Dict[str, Dict[str, Any]] = {}

# Session configured with timeout and custom User-Agent
session = requests.Session()
session.headers.update({
    'User-Agent': 'AppleMusicWidgetPlus/1.0.0 (https://github.com/28BEANS/apple-music-macOS-widget)'
})

def clean_search_query(text: str) -> str:
    """
    Normalizes search terms by stripping features and metadata noise.
    """
    if not text:
        return ""
    text = re.sub(r'\s*[\(\[]\s*feat(uring)?\..*?[\)\]]', '', text, flags=re.IGNORECASE)
    text = re.sub(r'\s*[\(\[]\s*with\s+.*?[\)\]]', '', text, flags=re.IGNORECASE)
    text = re.sub(r'\s*-\s*feat(uring)?\..*', '', text, flags=re.IGNORECASE)
    text = re.sub(r'\s*-\s*with\s+.*', '', text, flags=re.IGNORECASE)
    text = re.sub(r'\s*[\(\[]\s*(remastered|mono|stereo|radio edit|deluxe|version).*?[\)\]]', '', text, flags=re.IGNORECASE)
    return text.strip()

def fetch_from_lrclib(artist: str, track: str, album: str = "", duration: float = 0.0) -> Optional[Dict[str, Any]]:
    """
    Attempts to fetch lyrics from LRCLIB.
    """
    try:
        # 1. Exact match signature lookup
        if duration and duration > 0:
            url = "https://lrclib.net/api/get"
            response = session.get(url, params={
                "artist_name": artist,
                "track_name": track,
                "album_name": album or "",
                "duration": round(duration)
            }, timeout=8)
            if response.status_code == 200:
                data = response.json()
                if data.get("plainLyrics") or data.get("syncedLyrics"):
                    return {
                        "plainLyrics": data.get("plainLyrics") or "",
                        "syncedLyrics": data.get("syncedLyrics") or "",
                        "source": "LRCLIB"
                    }

        # 2. General search query search
        cleaned_query = f"{clean_search_query(artist)} {clean_search_query(track)}"
        search_url = "https://lrclib.net/api/search"
        response = session.get(search_url, params={"q": cleaned_query}, timeout=8)
        if response.status_code == 200:
            results = response.json()
            if isinstance(results, list) and len(results) > 0:
                for item in results:
                    if item.get("plainLyrics") or item.get("syncedLyrics"):
                        return {
                            "plainLyrics": item.get("plainLyrics") or "",
                            "syncedLyrics": item.get("syncedLyrics") or "",
                            "source": "LRCLIB"
                        }
    except Exception as e:
        print(f"LRCLIB fetch error: {e}")
    return None

def fetch_from_genius(artist: str, track: str) -> Optional[Dict[str, Any]]:
    """
    Searches Genius API and scrapes the song webpage HTML for lyrics.
    """
    api_key = os.getenv("GENIUS_API_KEY")
    if not api_key or api_key == "your_genius_api_key_here":
        print("Genius API Key is not set or using placeholder.")
        return None

    try:
        cleaned_artist = clean_search_query(artist)
        cleaned_track = clean_search_query(track)
        query = f"{cleaned_artist} {cleaned_track}"

        search_url = "https://api.genius.com/search"
        response = session.get(search_url, params={"q": query}, headers={
            "Authorization": f"Bearer {api_key}"
        }, timeout=8)

        if response.status_code != 200:
            return None

        data = response.json()
        hits = data.get("response", {}).get("hits", [])
        if not hits:
            return None

        # Extract first song result hit
        song_hit = next((hit for hit in hits if hit.get("type") == "song"), None)
        if not song_hit or not song_hit.get("result", {}).get("url"):
            return None

        song_url = song_hit["result"]["url"]
        
        # Request Genius lyrics page HTML
        page_response = session.get(song_url, timeout=8)
        if page_response.status_code != 200:
            return None

        # Parse lyrics text using BeautifulSoup scraper
        lyrics_text = extract_lyrics(page_response.text)
        if lyrics_text:
            return {
                "plainLyrics": lyrics_text,
                "syncedLyrics": "",
                "source": "Genius"
            }
    except Exception as e:
        print(f"Genius fetch error: {e}")
    return None

def get_lyrics(artist: str, track: str, album: str = "", duration: float = 0.0) -> Dict[str, Any]:
    """
    Main resolution handler with caching and fallback steps.
    """
    if not artist or not track:
        raise ValueError("Artist and track parameters are required.")

    cache_key = f"{artist.strip()} - {track.strip()}".lower()
    if cache_key in cache:
        print(f"[Cache Hit] {artist} - {track}")
        return cache[cache_key]

    print(f"[Resolution Init] {artist} - {track}")

    # Query LRCLIB first
    result = fetch_from_lrclib(artist, track, album, duration)

    # Fallback to Genius scraping
    if not result:
        print(f"[Genius Fallback] {artist} - {track}")
        result = fetch_from_genius(artist, track)

    if result:
        cache[cache_key] = result
        return result

    not_found = {
        "plainLyrics": "",
        "syncedLyrics": "",
        "source": "None"
    }
    cache[cache_key] = not_found
    return not_found

File: backend/lyrics/lyrics_service.js
import axios from 'axios';
import { extractLyrics } from './parser.js';

// Simple in-memory cache to prevent redundant API queries and scraping
const cache = new Map();

// HTTP Client configured with a reasonable timeout and User-Agent
const httpClient = axios.create({
  timeout: 8000,
  headers: {
    'User-Agent': 'AppleMusicWidgetPlus/1.0.0 (https://github.com/28BEANS/apple-music-macOS-widget)'
  }
});

/**
 * Normalizes song/artist names by stripping common noise like "feat.", "(Remastered)", etc.
 * @param {string} str - Raw input name.
 * @returns {string} Cleaned name.
 */
function cleanSearchQuery(str) {
  if (!str) return '';
  return str
    .replace(/\s*[\(\[]\s*feat(uring)?\..*?[\)\]]/gi, '')
    .replace(/\s*[\(\[]\s*with\s+.*?[\)\]]/gi, '')
    .replace(/\s*-\s*feat(uring)?\..*/gi, '')
    .replace(/\s*-\s*with\s+.*/gi, '')
    .replace(/\s*[\(\[]\s*(remastered|mono|stereo|radio edit|deluxe|version).*?[\)\]]/gi, '')
    .trim();
}

/**
 * Searches LRCLIB for lyrics.
 * @param {string} artist - Artist name.
 * @param {string} track - Track title.
 * @param {string} [album] - Album title.
 * @param {number} [duration] - Track duration in seconds.
 * @returns {Promise<object|null>} Lyric record if found, otherwise null.
 */
async function fetchFromLRCLIB(artist, track, album, duration) {
  try {
    // 1. Try exact match signature lookup if duration is available
    if (duration && duration > 0) {
      const url = `https://lrclib.net/api/get`;
      const response = await httpClient.get(url, {
        params: {
          artist_name: artist,
          track_name: track,
          album_name: album || '',
          duration: Math.round(duration)
        },
        validateStatus: (status) => status === 200 || status === 404
      });

      if (response.status === 200 && response.data && (response.data.plainLyrics || response.data.syncedLyrics)) {
        return {
          plainLyrics: response.data.plainLyrics || '',
          syncedLyrics: response.data.syncedLyrics || '',
          source: 'LRCLIB'
        };
      }
    }

    // 2. Fallback to general search query on LRCLIB
    const cleanedQuery = `${cleanSearchQuery(artist)} ${cleanSearchQuery(track)}`;
    const searchUrl = `https://lrclib.net/api/search`;
    const searchResponse = await httpClient.get(searchUrl, {
      params: { q: cleanedQuery }
    });

    if (searchResponse.status === 200 && Array.isArray(searchResponse.data) && searchResponse.data.length > 0) {
      // Find the first result with actual lyrics
      const bestMatch = searchResponse.data.find(item => item.plainLyrics || item.syncedLyrics);
      if (bestMatch) {
        return {
          plainLyrics: bestMatch.plainLyrics || '',
          syncedLyrics: bestMatch.syncedLyrics || '',
          source: 'LRCLIB'
        };
      }
    }
  } catch (error) {
    console.error('LRCLIB fetch error:', error.message);
  }
  return null;
}

/**
 * Resolves lyrics by searching Genius and scraping the song page HTML.
 * @param {string} artist - Artist name.
 * @param {string} track - Track title.
 * @returns {Promise<object|null>} Lyric record if found, otherwise null.
 */
async function fetchFromGenius(artist, track) {
  const apiKey = process.env.GENIUS_API_KEY;
  if (!apiKey || apiKey === 'your_genius_api_key_here') {
    console.warn('Genius API Key is not set or using placeholder.');
    return null;
  }

  try {
    const cleanedArtist = cleanSearchQuery(artist);
    const cleanedTrack = cleanSearchQuery(track);
    const query = `${cleanedArtist} ${cleanedTrack}`;

    // Search Genius API
    const searchUrl = `https://api.genius.com/search`;
    const searchResponse = await httpClient.get(searchUrl, {
      params: { q: query },
      headers: {
        'Authorization': `Bearer ${apiKey}`
      }
    });

    if (searchResponse.status !== 200 || !searchResponse.data.response) {
      return null;
    }

    const hits = searchResponse.data.response.hits;
    if (!hits || hits.length === 0) {
      return null;
    }

    // Select the first song type hit
    const songHit = hits.find(hit => hit.type === 'song');
    if (!songHit || !songHit.result || !songHit.result.url) {
      return null;
    }

    const songUrl = songHit.result.url;
    
    // Fetch Genius song webpage HTML
    const pageResponse = await httpClient.get(songUrl);
    if (pageResponse.status !== 200) {
      return null;
    }

    // Scrape/parse HTML using the cheerio parser
    const lyricsText = extractLyrics(pageResponse.data);
    if (lyricsText && lyricsText.length > 0) {
      return {
        plainLyrics: lyricsText,
        syncedLyrics: '', // Genius only has plain lyrics
        source: 'Genius'
      };
    }
  } catch (error) {
    console.error('Genius fetch error:', error.message);
  }
  return null;
}

/**
 * Coordinates lyric resolution using caching, LRCLIB, and Genius fallback.
 * @param {string} artist - Artist name.
 * @param {string} track - Track title.
 * @param {string} [album] - Album name.
 * @param {number} [duration] - Track duration in seconds.
 * @returns {Promise<object>} The resolved lyrics result.
 */
export async function getLyrics(artist, track, album = '', duration = 0) {
  if (!artist || !track) {
    throw new Error('Artist and track parameters are required.');
  }

  const cacheKey = `${artist.trim()} - ${track.trim()}`.toLowerCase();
  if (cache.has(cacheKey)) {
    console.log(`[Cache Hit] ${artist} - ${track}`);
    return cache.get(cacheKey);
  }

  console.log(`[Resolution Init] ${artist} - ${track}`);

  // Step 1: Query LRCLIB
  let result = await fetchFromLRCLIB(artist, track, album, duration);

  // Step 2: Fallback to Genius if LRCLIB returns empty
  if (!result) {
    console.log(`[Genius Fallback] ${artist} - ${track}`);
    result = await fetchFromGenius(artist, track);
  }

  // Step 3: Return result or final not found status
  if (result) {
    cache.set(cacheKey, result);
    return result;
  }

  const notFoundResult = {
    plainLyrics: '',
    syncedLyrics: '',
    source: 'None'
  };
  // Cache not found results as well to prevent spamming
  cache.set(cacheKey, notFoundResult);
  return notFoundResult;
}
